### PR TITLE
interpret: on a signed deref check, mention the right pointer in the error

### PR DESCRIPTION
**Pull request by RalfJung**
When a negative offset (like `ptr.offset(-10)`) goes out-of-bounds, we currently show an error saying that we expect the *resulting* pointer to be inbounds for 10 bytes. That's confusing, so this PR makes it so that instead we say that we expect the *original* pointer `ptr` to have 10 bytes *to the left*.

I also realized I can simplify the pointer arithmetic logic and handling of "staying inbounds of a target `usize`" quite a bit; the second commit does that.

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/memory.rs [hunk 1]
```diff
@@ -420,27 +422,26 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
+        Ok(())
     }
 
     /// Low-level helper function to check if a ptr is in-bounds and potentially return a reference
     /// to the allocation it points to. Supports both shared and mutable references, as the actual
-    /// checking is offloaded to a helper closure.
+    /// checking is offloaded to a helper closure. Supports signed sizes for checks "to the left" of
+    /// a pointer.
     ///
+    /// `size` must be between i64::MIN and u64::MAX.
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/memory.rs:437]: uh... it can't be `u64::MAX`, it's an `i64` now. did you want to use an `i128`?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/memory.rs:437]: Ah this is an outdated comment from when I thought we needed `i128`.

**Comment by oli-obk** [comment]
r=me with CI happy

**Comment by RalfJung** [comment]
Argh.
@bors r-

**Pull request merged**